### ISSUE TITLE
IUrlHelper.Link generation from action names

### ISSUE THREAD

**Issue by fragilethoughts** [aspnet/Mvc#7807]
Currently `IUrlHelper.Link` allows for generating absolute URLs when passing to it [route names which must be unique application-wide](https://docs.microsoft.com/en-us/aspnet/core/mvc/controllers/routing?view=aspnetcore-2.0#route-names). 

This method looks like this: `string Link(string routeName, object values)` and this is currently the only official way when implementing HATEOAS or any hypermedia format.

[**This doesn't work when someone works with generic controllers**](https://stackoverflow.com/questions/48633939/route-name-for-httpget-attribute-name-for-base-generic-controller-class-in-asp-n).

It would be nice if absolute URL could be generated using action name which should always work even with derived controllers with the same action method names.

Currently I solved it with extension defined in my project:

```csharp
public static class UrlHelperExtensions
{
    /// <summary>
    /// Generates an absolute URL for the specified <paramref name="actionName"/> and action
    /// <paramref name="values"/>, which contains the protocol (such as "http" or "https") and host name from the
    /// current request.
    /// </summary>
    /// <param name="actionName">The name of the action that is used to generate URL.</param>
    /// <param name="values">An object that contains action values.</param>
    /// <returns>The generated absolute URL.</returns>
    public static string LinkAction(this IUrlHelper urlHelper, string actionName, object values)
    {
        var httpContext = urlHelper.ActionContext.HttpContext;
        return urlHelper.Action(new UrlActionContext()
        {
            Action = actionName,
            Protocol = httpContext.Request.Scheme,
            Host = httpContext.Request.Host.ToUriComponent(),
            Values = values
        });
    }
}
```

**Comment by dougbu** [comment]
@enemyofthedawn 
> It would be nice if absolute URL could be generated using action name which should always work even with derived controllers with the same action method names.

Since you have a relatively simple fix for your scenario, what change in the framework would you like to see?

Side question:
> this is currently the only official way when implementing HATEOAS or any hypermedia format.

What about HATEOAS requires absolute URLs?

**Comment by konrad-san** [comment]
@dougbu I just thought it could be useful to have out of the box. Like there is `Link` for routes.

**Comment by dougbu** [comment]
`public static string Action(this IUrlHelper helper, string action, object values)` already exists and it returns a relative URL. So, are you looking for
``` c#
public static string ActionLink(this IUrlHelper helper, string action, object values)
```
(copying your fix) or
``` c#
public static string Action(this IUrlHelper helper, string action, object values, bool isAbsoluteUrl)
```
?

----
FYI we've discussed similar issues in the past. See #1966 and #2810 for example.

We're at parity with https://github.com/aspnet/AspNetWebStack already. MVC 5.x has `Content(string)` and `Link(string, object)` but nothing like the overload or new method requested here. That said, I can see the utility for SEO (for example).

**Comment by enemyofthedawn** [comment]
Yes I'm looking for something like what you're asking. A way to generate links (with http/https, domain and path) but using action names instead of route names. 

The extension method defined in my project solves this but this is just a suggestion that maybe it might be good to add this to the framework.

**Comment by mkArtakMSFT** [comment]
@glennc, will this be a priority for the 2.2 milestone, or should we move this to the backlog?

**Comment by mkArtakMSFT** [comment]
@enemyofthedawn, would you like to submit a PR for this? We'd be happy to take it.

**Comment by enemyofthedawn** [comment]
@mkArtakMSFT Sure! Should it be named `LinkAction` or `ActionLink` ? Wouldn't it be better to rename `Link` to `RouteLink` and add `ActionLink` too or would it be a breaking change?

Just like it is named in HtmlHelper:

https://github.com/aspnet/Mvc/blob/a736441ca5ac7524f4ed0d9c33370becb523c778/src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/HtmlHelper.cs#L594

and

https://github.com/aspnet/Mvc/blob/a736441ca5ac7524f4ed0d9c33370becb523c778/src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/HtmlHelper.cs#L194

**Comment by dougbu** [comment]
> Should it be named `LinkAction` or `ActionLink`?

`ActionLink` is fine and aligns with `Link`.

> a breaking change?

Do not rename the existing methods because almost any change to `IUrlHelper` would be a breaking change. Instead, add something like what you described above to `UrlHelperExtensions`.

**Comment by enemyofthedawn** [comment]
> Instead, add something like what you described above to UrlHelperExtensions.

do you mean I shouldn't add it to `UrlHelper.cs` and `IUrlHelper.cs` and add it as an extension in a separate file?

**Comment by dougbu** [comment]
> add it as an extension in a separate file?

Exactly though `UrlHelperExtensions` already exists.

**Comment by SonicGD** [comment]
Hello. Because old PR was closed, i've reimplemented this feature with required fixes.

**Comment by blowdart** [comment]
We're relying on user input here again, in the form of the host header. This has created problems in the past. We've [rejected](https://github.com/aspnet/Mvc/issues/8279#issuecomment-414396069) this type of PR before and the same reasons apply here, so I'd view this, as I viewed the previous issue, as a won't fix.

**Comment by mkArtakMSFT** [comment]
Thanks for your effort, @SonicGD.
We're for now closing this issue, per @blowdart's comment.
We're going to address this as part of https://github.com/aspnet/Mvc/issues/8534 planned for 3.0 milestone.

**Comment by SonicGD** [comment]
Ok, no problem :)

**Comment by mkArtakMSFT** [comment]
Reopening this as seems that all the required bits for this to happen are now available.

**Comment by pranavkm** [comment]
cc @danroth27 could you suggest a name for this?

**Comment by poke** [comment]
(Reposting my comment from the PR here for visibility)

What about having a `Url.Absolute()` instead that would turn a relative URL into an absolute one?

So this would become `Url.Absolute(Url.Action(…))` but at the same time we also had `Url.Absolute(Url.Route(…))` or `Url.Absolute(Url.Page(…))` etc. without breaking compatibility and without ending up with lots of new methods.

And we could also use that implementation directly for #8279 then.
